File: backend/app/models/match_event.py
import enum

from sqlalchemy import Enum as SQLAlchemyEnum
from sqlalchemy import ForeignKey, Index, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, TimestampMixin


class EventType(str, enum.Enum):
    GOAL = "goal"
    OWN_GOAL = "own_goal"
    PENALTY_GOAL = "penalty_goal"
    PENALTY_MISS = "penalty_miss"
    YELLOW_CARD = "yellow_card"
    RED_CARD = "red_card"
    SUBSTITUTION = "substitution"
    VAR = "var"
    OTHER = "other"


class MatchEvent(TimestampMixin, Base):
    __tablename__ = "match_events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)

    match_id: Mapped[int] = mapped_column(ForeignKey("matches.id"), nullable=False)
    team_id: Mapped[int] = mapped_column(ForeignKey("teams.id"), nullable=False)

    player_id: Mapped[int | None] = mapped_column(ForeignKey("players.id"), nullable=True)
    secondary_player_id: Mapped[int | None] = mapped_column(ForeignKey("players.id"), nullable=True)

    player_external_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    secondary_player_external_id: Mapped[int | None] = mapped_column(Integer, nullable=True)

    player_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    secondary_player_name: Mapped[str | None] = mapped_column(String(100), nullable=True)

    event_type: Mapped[EventType] = mapped_column(
        SQLAlchemyEnum(
            EventType,
            values_callable=lambda enum_cls: [member.value for member in enum_cls],
            name="event_type_enum",
        ),
        nullable=False,
    )

    minute: Mapped[int] = mapped_column(Integer, nullable=False)
    extra_minute: Mapped[int | None] = mapped_column(Integer, nullable=True)

    detail: Mapped[str | None] = mapped_column(String(100), nullable=True)
    comments: Mapped[str | None] = mapped_column(String(255), nullable=True)

    team = relationship("Team")
    player = relationship("Player", foreign_keys=[player_id])
    secondary_player = relationship("Player", foreign_keys=[secondary_player_id])

    __table_args__ = (
        Index("ix_match_events_match_id", "match_id"),
        Index("ix_match_events_player_id", "player_id"),
        Index("ix_match_events_team_id", "team_id"),
    )

File: frontend/src/types/matchEvent.ts
import type { LucideIcon } from 'lucide-react';

import type { ResponseMetadata } from './metadata';
import type { Player } from './player';
import type { Team } from './team';

export type EventType =
  | 'goal'
  | 'own_goal'
  | 'penalty_goal'
  | 'penalty_miss'
  | 'yellow_card'
  | 'red_card'
  | 'substitution'
  | 'var'
  | 'other';

export interface MatchEvent {
  team: Team;
  player: Player | null;
  secondary_player: Player | null;
  player_name: string | null;
  secondary_player_name: string | null;
  player_external_id: number | null;
  secondary_player_external_id: number | null;
  event_type: EventType;
  minute: number;
  extra_minute: number | null;
  detail: string | null;
  comments: string | null;
}

export interface MatchEventResponse {
  data: MatchEvent[];
  metadata: ResponseMetadata;
}

export interface MatchEventsOptions {
  match_id: number;
}

export interface EventConfig {
  title: string;
  icon: LucideIcon;
  iconClassName: string;
  cardClassName: string;
}

export interface TimelineMarkerConfig {
  minute: number;
  label: string;
  order?: number;
}

export type TimelineItem =
  | {
      type: 'event';
      minute: number;
      order: number;
      event: MatchEvent;
      score: string;
    }
  | {
      type: 'marker';
      minute: number;
      order: number;
      label: string;
    };
